docs/design/metrics-best-practices.md

```diff
@@ -0,0 +1,70 @@
+# Kube-State-Metrics - Timeseries best practices
+
+---
+
+Author: Manuel Rüger (<manuel@rueg.eu>)
+
+Date: October 17th 2024
+
+---
+
+## Introduction
+
+Kube-State-Metrics' goal is to provide insights into the state of Kubernetes objects by exposing them as metrics.
+This document provides guidelines with the goal to create a good user experience when using these metrics.
+
+Please be aware that this document is introduced in a later stage of the project and there might be metrics that do not follow these best practices.
+Feel encouraged to report these metrics and provide a pull request to improve them.
+
+## General best practices
+
+We follow [Prometheus](https://prometheus.io/docs/practices/naming/) best practices in terms of naming and labeloing.
+
+## Best practices for kube-state-metrics
+
+### Avoid pre-computation
+
+kube-state-metrics should expose metrics on an individual object level and avoid any sort of pre-computation unless it is required due to for example high cardinality on objects.
+By exposing raw metrics instead of counters, kube-state-metrics allows the user to have full control on how they want to use the metrics.
+
+### Static object properties
+
+An object usually has a stable set of properties that do not change during its lifecycle in Kubernetes.
+This includes properties like name, namespace, uid etc.
+It is a good practice to group those together into an `_info` metric.
+
+### Dynamic object properties
+
+An object can also have a dynamic set of properties, which are usually part of the status field.
+These change during the lifecycle of the object.
+For example a pod can be in different states like "Pending", "Running" etc.
+These should be part of a new metric that includes labels that identify the object as well as the dynamic property.
+
+### Linked properties
+
+If an object contains a substructure that links multiple properties together (e.g. endpoint address and port), those should be reported in the same metric.
+
+### Optional properties
+
+Some Kubernetes objects have optional fields. In case there is an optional value, it is better to not expose the label at all instead of exposing a "nil" value or an empty string.
```

Comment:
Hmm, I'm not sure about this one.

Prometheus metric families should have consistent labels.

For example, this is considered invalid.
```
my_metric{a="foo",b="bar"} 1
my_metric{b="baz"}1
```

In the above case, you want to expose an empty string label so the metric family is consistent.
```
my_metric{a="foo",b="bar"} 1
my_metric{a="",b="baz"}1
```

Reply:
That's good to know! I'll fix it (and will add some fixes to the next release of ksm).
Is there any upstream documentation on this that suggests labels to be empty instead of leaving them out so I can reference them in here?

Reply:
I thought we had an exmaple in the OpenMetrics exposition spec, but I can't find it right now. I'll see if I can find the docs on this.

Reply:
The technicality here can be split into two considerations:

- The lint enforcement mentioned above comes about as a result of the [registry machinery](https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L310-L326).
- KSM [does not register metrics](https://github.com/kubernetes/kube-state-metrics/blob/0a86b24b08431c597519e19937b07162247035e3/pkg/customresourcestate/registry_factory.go#L219) in a registry since none of these are structured (except KSM's own meta-metrics, but piped-out the in the same [non-proto manner](https://github.com/kubernetes/kube-state-metrics/blob/166921b797100dadf51a153c8b932aab73dab4ff/pkg/metricshandler/metrics_handler.go#L202)).

KSM bypasses this trait by design (a `promtool check metrics` will succeed for the invalid example above).

<img width="805" alt="Screenshot 2024-10-24 at 1 01 59 PM" src="https://github.com/user-attachments/assets/7f6e74b4-41df-4b8b-a5f0-3815b3463e4b">

That being said, it's always good to be consistent with the ecosystem, given we can incorporate a solution that ensures this in the CI.